pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -228,6 +237,171 @@ LOOP:
+						q.logger.Error().Err(err).Msg("error scanning and leasing shards")
+						return
+					}
+				}
+			}()
+		case <-leaseTick.C:
+			for _, s := range q.shardLeases {
+				// Attempt to lease all ASAP, even if the backing store is single threaded.
+				go func(ls leasedShard) {
+					nextLeaseID, err := q.renewShardLease(ctx, &ls.Shard, ShardLeaseTime, ls.Lease)
+					if err != nil {
+						q.logger.Error().Err(err).Msg("error renewing shard lease")
+						return
+					}
+					// Update the lease ID so that we have this stored appropriately for
+					// the next renewal.
+					q.addLeasedShard(ctx, &ls.Shard, *nextLeaseID)
+				}(s)
+			}
+		}
+	}
+}
+
+func (q *queue) scanShards(ctx context.Context) (retry bool, err error) {
+	// TODO: Make instances of *queue register worker information when calling
+	//       Run().
+	//       Fetch this information, and correctly assign workers to shard maps
+	//       based on the distribution of items in the queue here.  This lets
+	//       us oversubscribe appropriately.
+
+	shardMap, err := q.getShards(ctx)
+	if err != nil {
+		q.logger.Error().Err(err).Msg("error fetching shards")
+		return
+	}
+	shards, err := q.filterShards(ctx, shardMap)
+	if err != nil {
+		q.logger.Error().Err(err).Msg("error filtering shards")
+		return
+	}
+	if len(shards) == 0 {
+		return
+	}
+
+	for _, shard := range shards {
+		leaseID, err := q.leaseShard(ctx, shard, ShardLeaseTime, len(shard.Leases))
+		if err == nil {
+			q.addLeasedShard(ctx, shard, *leaseID)
+		}
+		switch err {
+		case errShardNotFound:
+			// This is okay;  the shard was removed when trying to lease
+			continue
+		case errShardIndexLeased:
+			// This is okay;  another worker grabbed the lease.  No need to retry
+			// as another worker grabbed this.
+			continue
+		case errShardIndexInvalid:
+			// A lease expired while trying to lease — try again.
+			retry = true
+		default:
+			return true, err
+		}
+	}
+
+	return retry, nil
+}
+
+func (q *queue) addLeasedShard(ctx context.Context, shard *QueueShard, lease ulid.ULID) {
+	for i, n := range q.shardLeases {
+		if n.Shard.Name == shard.Name {
+			// Updated in place.
+			q.shardLeases[i] = leasedShard{
+				Lease: lease,
+				Shard: *shard,
+			}
+			return
+		}
+	}
+	// Not updated in place, so add to the list and return.
+	q.shardLeases = append(q.shardLeases, leasedShard{
+		Lease: lease,
+		Shard: *shard,
+	})
+}
+
+// filterShards filters shards during assignment, removing any shards that this worker
+// has already leased;  any shards that have already had their leasing requirements met;
+// and priority shuffles shards to lease in a non-deterministic (but prioritized) order.
+//
+// The returned shards are safe to be leased, and should be attempted in-order.
+func (q *queue) filterShards(ctx context.Context, shards map[string]*QueueShard) ([]*QueueShard, error) {
+	if len(shards) == 0 {
+		return nil, nil
+	}
+
+	for _, v := range q.shardLeases {
+		delete(shards, v.Shard.Name)
```

Comment:
same here regarding locking.

Reply:
This one is lock-free - it's passed in data from a single goroutine.